document-policy-explainer.md

```diff
@@ -0,0 +1,341 @@
+of synchronous APIs, or use of APIs that fall under iframe sandboxing,
+traditionally.
+
+That header doesn't impose any restrictions on content which the document embeds
+at all, and in the base case, every document is free to set its own policy
+independently.
+
+(In a minute, we'll talk about required policies, though -- and not conforming
+to a required policy might impact your page's ability to be loaded, but in
+general, you're free to set your policy any way you like)
+
+Policies controllable with this mechanism include things like document
+configuration, and sandboxing. Each of them can be named with a unique string.
+
+Examples:
+
+* `image-compression`
+* `frame-loading`
+* `vertical-scroll`
+* `popups`
+* `document-write`
+* `scripts`
+
+Some policies are boolean switches, and can be turned off with the syntax
+"`no-<feature-name>`":
+
+* `no-popups`
+* `no-scripts`
+* `no-document-write`
+
+Others take parameters, whose types and names are specific to each policy:
+
+* `image-compression;min-bpp=2`
+* `frame-loading;lazy`
+
+All of the directives can be combined with commas into a single header. In
+Structured-header-speak, this is a parameterized list.
+
+### Requiring a policy in your embedded content
+
+It is possible to enforce that content you embed conforms to a specific policy
+as well. This can be done in two ways -- either through a different response
+header: `Require-Document-Policy`, which affects every iframe on your page, or
+by using a "`policy`" attribute on specific frames, which target just those
+frames.
+
+`Require-Document-Policy` doesn't affect the document that it is returned with,
+just the content that it embeds. It can be used in conjunction with the
+`Document-Policy` header, though.
+
+When you set a required document policy, most features will have to be
+acknowledged by the document being embedded, as otherwise it can be possible to
+affect the content on that page in unexpected ways. Any policies which require
+acknowledgement will be advertised with a `Sec-Required-Document-Policy` request
+header.
+
+If the content which is returned doesn't have a `Document-Policy` header which
+is at least as strict as the `Sec-Required-Document-Policy`, then it will not be
+loaded into the frame, and a network error will result instead.
+
+The `Sec-Required-Document-Policy` header will be present for all requests for
+documents in a frame, at every nesting level. If the framed document adds its
+own `Require-Document-Policy header`, or uses the policy attribute on its own
+iframes, then the requirements will be combined in deeper requests, to ensure
+that all content meets all of the requirements being imposed on it.
+
+### Extending the Iframe Sandboxing Mechanism
+
+All existing sandbox features can be defined as document policies as well. The
+iframe `sandbox` attribute will disable all of them by default, but they can be
+turned back on either with the `policy` attribute or the `sandbox` attribute.
```

Comment:
What would it look like to turn sandbox flag back off with the `policy` attribute? Would it be `<iframe sandbox policy="forms">`, as an equivalent to `<iframe sandbox="allow-forms">`?

Reply:
I think that would be the syntax, if we can do it. I'm worried about the compat implications of allowing *both* `<iframs sandbox="allow-forms">` and `<iframe sandbox policy="forms">` though.

Reply:
Do you mean the risk of people adopting `<iframe sandbox policy="forms">` before it works in all browsers, and ending up not having the behavior they want in some browsers? There's also the potential confusion of having two equivalent ways to express two things. What's the strongest argument for having this new form? Tradeoffs :)